ios-sockets/SwiftWebSockets/SwiftWebSockets/Networking/NativeWebSocket.swift

```diff
@@ -106,19 +124,33 @@ class NativeWebSocket: NSObject, WebSocketConnection, URLSessionWebSocketDelegat
         }
     }
     
-    func ping(with frequency: TimeInterval = 25.0) {
-        pingTimer = Timer.scheduledTimer(withTimeInterval: frequency, repeats: true) { [weak self] _ in
+    func ping(interval: TimeInterval = 25.0) {
+        pingTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
             guard let self = self else { return }
-            self.webSocketTask.sendPing { error in
-                if let error = error {
-                    self.delegate?.onError(connection: self, error: error)
-                }
+            self.ping()
+        }
+    }
+
+    func ping() {
+        self.webSocketTask.sendPing { error in
+            if let error = error {
+                self.delegate?.webSocketDidReceiveError(connection: self, error: error)
             }
         }
     }
     
-    func disconnect() {
-        webSocketTask.cancel(with: .normalClosure, reason: nil)
+    func disconnect(closeCode: NWProtocolWebSocket.CloseCode) {
+        var webSocketTaskCloseCode: URLSessionWebSocketTask.CloseCode!
+        switch closeCode {
+        case .protocolCode(let definedCode):
+            webSocketTaskCloseCode = URLSessionWebSocketTask.CloseCode(rawValue: Int(definedCode.rawValue))
+        case .applicationCode, .privateCode:
```

Comment:
Interesting! So this is how you're able to cover all the close codes for your use case.  Is there a way to be more flexible with how to handle private codes for others using the disconnect logic?

Reply:
So, this is the motivation for creating `NWWebSocket` altogether - you'll notice that `NWProtocolWebSocket.CloseCode` covers protocol, application and private closure code ranges. `URLSessionWebSocketTask.CloseCode` however, only covers the protocol closure codes (i.e. those also defined in `NWProtocolWebSocket.CloseCode.Defined`).

So the flexibility for my use case is actually all inside `NWWebSocket`'s `disconnect(…)` method. The mapping you've pointed to is because `URLSessionWebSocketTask` cannot have any custom closure codes passed via it's `cancel(with closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?)` method.

Reply:
That makes perfect sense, thank you!